src/prefect/runtime/flow_run.py

```diff
@@ -29,14 +30,36 @@ def __getattr__(name: str) -> Any:
 
         from prefect.runtime.flow_run import id
     """
-    env_key = f"PREFECT__RUNTIME__FLOW_RUN__{name.upper()}"
-    if env_key in os.environ:
-        return os.environ[env_key]
+
     func = FIELDS.get(name)
-    if func is None:
-        raise AttributeError(f"{__name__} has no attribute {name!r}")
+
+    # if `name` is an attribute but it is mocked through environment variable, the mocked type will be str,
+    # which might be different from original one. For consistency, cast env var to the same type
+    env_key = f"PREFECT__RUNTIME__FLOW_RUN__{name.upper()}"
+
+    if func is not None:
+        real_value = func()
+        if env_key in os.environ:
+            mocked_value = os.environ[env_key]
+            # cast `mocked_value` to the same type than `real_value`
+            if isinstance(real_value, bool):
+                return bool(mocked_value)
+            elif isinstance(real_value, int):
+                return int(mocked_value)
+            elif isinstance(real_value, float):
+                return float(mocked_value)
+            elif isinstance(real_value, pendulum.DateTime):
+                return pendulum.instance(dateparser.parse(mocked_value))
+            else:
+                # default str
+                return mocked_value
+        else:
+            return real_value
     else:
-        return func()
+        if env_key in os.environ:
+            return os.environ[env_key]
+        else:
+            raise AttributeError(f"{__name__} has no attribute {name!r}")
```

Comment:
It might be more readable to check `if func is None` first-- that way we can quickly `return` or `raise`. With that change, we can then make `if env_key in os.environ` (where `func is not None`) a top-level/outermost condition.

Reply:
Totally agree!